### PR TITLE
CTCP-3453: Consolidate business validation & and toggleable business validation rule

### PR DESCRIPTION
**Pull request by drnaylor-oc**
# What is this PR about?

This PR started life as a way to consolidate the business rule validation into one flow for both XML and Json formats, to make it easier to add business rules in one place in the future.

This was in anticipation of CTCP-3453, which got clarified faster than I thought it would. It has been added here - the ticket was always going to come into sprint anyway.

# What does it do?

The old business validation code has been removed and rewritten to take advantage of Akka's streaming, and to allow us to process the business rules at the same time as the schema validation. The output is the same, but the logic is now not duplicated (and state has been removed from our code).

# How does it work?

There is a little bit of deferred calculation in here, but the main points are:

* XML/Json specific processes for business rules have been moved to a `MessageFormat` object
* Business rules are now in their own service, and returns a pre-materialised `Flow` and associated `Future` that the flow will populate when the `Flow` is attached to a stream and executed.
* That flow is attached to the request stream and then executed in the schema validation services.
* When that flow completes, if there were no schema errors, the future is inspected for any errors.

The streaming still happens from a file, at least for now. We can do a bit more work to resolve that later, though we need to be careful to drain the request if an incorrect message type is provided.

# How does the Flow work?

The flow forks the stream before we send it for schema validation, and parses the stream as XML or Json, before feeding the various rules that look for the conditions. The results are sent onto a sink that reads the first error, if any, which is what the `Future` reports.

# Why is there a new configuration option?

An incoming requirement is that the `messageRecipient` field **must** be either "NTA.GB" or "NTA.XI", and must match the country of the appropriate office we're targeting. However, we don't want to enforce it on third parties all of a sudden, we need to give them a bit of notice. This allows for our testing to proceed, but for us to turn it on in the sandbox at a later time.

# Why is it off by default now?

So we can get it into the codebase as the consolidated error handling is good to have, and allows QA to build their scenarios under less pressure.

### CODE REVIEW

File: app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala [hunk 1]
```diff
@@ -49,226 +41,65 @@ import scala.concurrent.ExecutionContext
...
-                      errorBuffer += XmlSchemaValidationError.fromSaxParseException(error)
-                  })
+      parsersByType(messageType).map {
+        saxParser =>
+          Using(source.runWith(StreamConverters.asInputStream(20.seconds))) {
+            xmlInput =>
+              val parser = saxParser.newSAXParser.getParser
 
-                  val inputSource = new InputSource(xmlInput)
+              val errorBuffer: mutable.ListBuffer[XmlSchemaValidationError] =
```
Comment (derekspalmer) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala:67]: Ok. I see why it's mutable.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala [hunk 2]
```diff
@@ -77,130 +70,40 @@ trait JsonValidationService extends ValidationService
...
-  //and is one of the defined departure values, it returns the string "OfficeOfDeparture".
-  //If the message type is a subtype of ArrivalMessageType and is one of the defined arrival values, it returns the string "OfficeOfDestinationActual". For all other cases, it returns a None.
-  def checkMessageType(rootTag: String): Option[MessageType] = {
-    val messageType = MessageType.values.find(_.rootNode.equalsIgnoreCase(rootTag.split(":")(1)))
-    messageType
-  }
-
-  def validateJson(source: Source[ByteString, _], schemaValidator: JsonSchema)(implicit materializer: Materializer): Try[Set[ValidationMessage]] =
+  private def validateJson(source: Source[ByteString, _], schemaValidator: JsonSchema)(implicit materializer: Materializer): Try[Set[ValidationMessage]] =
     Using(source.runWith(StreamConverters.asInputStream(20.seconds))) {
```
Comment (derekspalmer) [inline app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala:101]: I know it's not part of this, but do you remember why we used 20 seconds?

**Review by derekspalmer — APPROVED** [review]
LGTM.

File: app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala (hunk 2 above)
Comment (drnaylor-oc) [reply app/uk/gov/hmrc/transitmovementsvalidator/services/JsonValidationService.scala:101]: That's the standard timeout across HMRC services, including CTC, so if this takes longer than 20 seconds, the trader would have timed out anyway.

**Pull request merged**